redis/src/commands/cluster_scan.rs

```diff
@@ -0,0 +1,639 @@
+        if let Some(match_pattern) = match_pattern {
+            scan_command.arg("MATCH").arg(match_pattern);
+        }
+        if let Some(count) = count {
+            scan_command.arg("COUNT").arg(count);
+        }
+        if let Some(object_type) = object_type {
+            scan_command.arg("TYPE").arg(object_type);
+        }
+        let scan_result = self
+            .route_command(scan_command, &scan_state.node_in_scan)
+            .await;
+        if scan_result.is_err() {
+            return Err(RedisError::from((
+                ErrorKind::IoError,
+                "Failed to route command",
+            )));
+        }
+        let (new_cursor, new_keys): (u64, Vec<Value>) =
+            from_redis_value(&scan_result.unwrap()).unwrap();
+        // If the cursor is 0, meaning we finished scanning the node
+        if new_cursor == 0 {
+            scan_state = scan_state.update_scan_state_and_get_next_node(self).await;
+            // If the node is empty, meaning we finished scanning all the nodes
+            if scan_state.node_in_scan.is_empty() {
+                return Ok((ScanStateWrapper::new(), new_keys));
+            }
+        }
+        scan_state = ScanState::create(
+            new_cursor,
+            scan_state.scanned_slots_map,
+            scan_state.slots_in_scan_map,
+            scan_state.node_in_scan,
+            scan_state.node_epoch,
+        );
+        Ok((ScanStateWrapper::from(&scan_state), new_keys))
+    }
+}
+
+/// Represents the state of a scan operation in a Redis cluster.
+///
+/// This struct holds information about the current scan state, including the cursor position,
+/// the scanned slots map, the slots in the scan map, the node being scanned, and the node's epoch.
+impl ScanState {
+    /// Create a new instance of ScanState.
+    ///
+    /// # Arguments
+    ///
+    /// * `cursor` - The cursor position.
+    /// * `scanned_slots_map` - The scanned slots map.
+    /// * `slots_in_scan_map` - The slots in the scan map.
+    /// * `node_in_scan` - The node being scanned.
+    /// * `node_epoch` - The epoch of the node being scanned.
+    ///
+    /// # Returns
+    ///
+    /// A new instance of ScanState.
+    pub fn create(
+        cursor: u64,
+        scanned_slots_map: SlotsBitsArray,
+        slots_in_scan_map: SlotsBitsArray,
+        node_in_scan: String,
+        node_epoch: u64,
+    ) -> Self {
+        Self {
+            cursor,
+            scanned_slots_map,
+            slots_in_scan_map,
+            node_in_scan,
+            node_epoch,
+        }
+    }
+
+    /// Check if the slots that being scanned has changed and the node has some slots missing.
+    ///
+    /// # Arguments
+    ///
+    /// * `connection` - The connection to the Redis cluster.
+    ///
+    /// # Returns
+    ///
+    /// `true` if the slots of the node were changed, `false` otherwise.
+    pub(crate) async fn check_if_node_slots_changed<C: ClusterInScan + ?Sized>(
```

Comment:
This function doesn't check if slots were removed from the node. 
if slots under scan are [0, 7, 10, 11] and the current slots are [0, 7, 10] the function will return false - that there wasn't any change in slots. even though the function name implies that it should return true. If that's ok from the scan algorithm POV, it should be documented, and the function name should reflect what the function really does. 


I couldn't find tests for all scenarios for this function - slots were removed and the expected output, slots were added, slots are the same. Please separate the scenarios for different test functions where each should state what it checks.